### PR TITLE
[charts] Add sparkline

### LINKED ISSUE #2902: Add Sparkline component
- [x] I have searched the [issues](https://github.com/mui-org/material-ui/issues) of this repository and believe that this is not a duplicate.

## Summary 💡

Sparklines are commonly used in enterprise applications to display simple metrics.
## Examples 🌈

A line chart:

Capture d’écran 2020-09-04 à 13 07 44

They can be of different shapes, like bar chart.

## Motivation 🔦

It's meant as a simplified chart. You would use it if you have live data or a simple set of points you want to display in a dense area. They can be embedded in tables, headlines, or chunks of text. It's perfect when you don't want to introduce a large chart.

## Benchmark

https://trello.com/c/FV7kGKxE/2177-sparkline-component

### PR DESCRIPTION
**Pull request by alexfauquette**
Fix #2902 

@LukasTy The spark line is basically a rewriting of `LineCharts` or `BarCharts` with less options/components

I tried to reorganize of props that are passed to the component such that it feels more natural. For example, I assume it can only have one series, so the `series` prop does not exist and you directly use some `data` props.
Plus some helpers to simplify adding/removing basics interactions.

About the DX I tried to create the component I would like to use if I had an array of data to plot in a small part of an app, but I'm not that sure of the result

### CODE REVIEW

File: docs/data/charts/sparkline/sparkline.md [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+---
+
+# Charts - Sparkline
+
+<p class="description">Sparkline charts can provide an overview of data trends.</p>
+
+## Basics
+
+The `<SparklineCharts />` requires only the `data` props which is an array of numbers.
+You can also choose between spark ann line plot with `plotType="bar"`.
```
Comment (joserodolfofreitas) [inline docs/data/charts/sparkline/sparkline.md:12]:
```suggestion
You can also switch from line to a bar plot with `plotType="bar"`.
```

File: docs/data/charts/sparkline/sparkline.md [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+{{"demo": "BasicSparkLine.js"}}
+
+## Line customization
+
+You can fill the area below the sparkline curve with `area` props.
+You can also modify the interpolation with `curve` props, as explained in the [line charts docs](/x/react-charts/lines/#interpolation).
```
Comment (joserodolfofreitas) [inline docs/data/charts/sparkline/sparkline.md:19]:
```suggestion
To modify the interpolation with `curve` props, follow the instructions in the [line charts docs](/x/react-charts/lines/#interpolation).
```

File: docs/data/charts/sparkline/sparkline.md [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+{{"demo": "AreaSparkLine.js"}}
+
+## Interaction
+
+Compared to line and bar charts, the spark-line chart has some additional props to simplify interaction configuration.
```
Comment (joserodolfofreitas) [inline docs/data/charts/sparkline/sparkline.md:25]:
```suggestion
Compared to line and bar charts, the sparkline chart has some additional props to simplify interaction configuration.
```

File: docs/data/charts/sparkline/sparkline.md [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+You can use `showTooltip` and `showHighlight` to add the default tooltip and highlight to your spark-line component.
```
Comment (joserodolfofreitas) [inline docs/data/charts/sparkline/sparkline.md:26]:
```suggestion
You can use `showTooltip` and `showHighlight` to display the default tooltip and highlight in your sparkline.
```

File: docs/data/charts/sparkline/sparkline.md [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+
+Those are helpers.
+If you need more advanced customization, you can still provide the tooltip and highlight props described on the [dedicated docs page](/x/react-charts/tooltip/).
```
Comment (joserodolfofreitas) [inline docs/data/charts/sparkline/sparkline.md:29]:
```suggestion
If you need more advanced customization, you can provide custom props for `tooltip` and `highlight` as described in the [Tooltip page](/x/react-charts/tooltip/).
```

File: docs/data/charts/sparkline/sparkline.md [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+{{"demo": "BasicSparkLineCustomization.js"}}
+
+## Axis management
+
+By default, the sparkline will set x values to 0, 1, 2, ... and hide it in the tooltip.
```
Comment (joserodolfofreitas) [inline docs/data/charts/sparkline/sparkline.md:35]:
```suggestion
By default, the sparkline assigns `xAxis` values as an ascending integer sequence starting from 0 (0, 1, 2,...).  These values are, in this case, hidden in the tooltip.
```

File: docs/data/charts/sparkline/sparkline.md [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+You can override this behavior if your data are not evenly distributed, or if their value matter.
```
Comment (joserodolfofreitas) [inline docs/data/charts/sparkline/sparkline.md:36]:
```suggestion
You can override this behavior if your data are not evenly distributed, or if you need to label them.
```

Should we also push a bit to showcase an evenly distributed data?

image

File: docs/data/charts/tooltip/tooltip.md [hunk 8]
```diff
@@ -65,6 +67,18 @@ Here is a demo with:
 
 {{"demo": "Formatting.js"}}
 
+### Hiding values
+
+You can hide the axis value with `tooltipHiden` in the `xAxis` props.
```
Comment (joserodolfofreitas) [inline docs/data/charts/tooltip/tooltip.md:72]:
```suggestion
You can hide the axis value with `tooltipHidden` in the `xAxis` props.
```

File: docs/data/charts/tooltip/tooltip.md [hunk 9]
```diff
@@ -65,6 +67,18 @@ Here is a demo with:
...
+It will remove the header showing the x-axis value from the tooltip.
+
+```jsx
+<LineChart
+  // ...
+  xAxis={[{ data: [ ... ], tooltipHiden: true }]}
```
Comment (joserodolfofreitas) [inline docs/data/charts/tooltip/tooltip.md:78]:
```suggestion
  xAxis={[{ data: [ ... ], tooltipHidden: true }]}
```

**Review by joserodolfofreitas — COMMENTED** [review]
Awesome abstraction for the Sparkline! Congrats!

File: docs/data/charts/sparkline/sparkline.md [hunk 10]
```diff
@@ -0,0 +1,46 @@
+---
+title: Charts - Sparkline
+---
+
+# Charts - Sparkline
+
+<p class="description">Sparkline charts can provide an overview of data trends.</p>
+
+## Basics
+
```
Comment (joserodolfofreitas) [inline docs/data/charts/sparkline/sparkline.md:10]: I feel we could add a brief explanation for it.
```suggestion

A sparkline is a small chart drawn without axes or coordinates to present the general shape of a variation in a simplified way.
```

**Comment by flaviendelangle** [comment]
In term of bundle size, do you have an idea how much lighter it would be to separated bar and line sparklines ?
Bundle size is probably a lot  more important for these small less-important charts.

**Comment by alexfauquette** [comment]
Here is the bundle analyze of a `creat-react-app` with a sparkline and the zoom on the charts package

image
image

We could save 7kB with such a split. 

We could also be more aggressive and put the tooltip as a slot to provide to remove also 5kB from `ChartsTooltip` and `useAxisEvents`

Later we could also remove the formatters/extremumGetters function since each plot type adds around 1kB

I'm not used to bundle size issues, so I've no idea about magnitude orders

File: packages/x-charts/src/SparkLineChart/SparkLineChart.tsx [hunk 11]
```diff
@@ -0,0 +1,267 @@
...
+  xAxis?: MakeOptional<AxisConfig, 'id'>;
+  tooltip?: ChartsTooltipProps;
+  axisHighlight?: ChartsAxisHighlightProps;
+  /**
+   * Type of plot used.
+   * @default 'line'
+   */
+  plotType?: 'line' | 'bar';
+  /**
+   * Data to plot
```
Comment (LukasTy) [inline packages/x-charts/src/SparkLineChart/SparkLineChart.tsx:39]:
```suggestion
   * Data to plot.
```

**Review by LukasTy — APPROVED** [review]
Great work! 👍 
Works and looks very nice. 💯 
One tiny nitpick could be a suggestion to include smaller examples, as usually you'd expect sparklines tucked away somewhere tight. 🤔 
But, on the other hand, it could be a future additional reference, once we have a data grid example, just back-link it here. 🙈

**Comment by flaviendelangle** [comment]
> We could save 7kB with such a split.

7 out of 31 is quite a nice improvement
I honestly don't know how far we want to go with bundle size on the charts

**Comment by alexfauquette** [comment]
> 7 out of 31 is quite a nice improvement

Depends on which use case you are considering

If they use bar/line chart somewhere else in their application, the kB saved by this modification goes down to 0 since 

If sparkling is the only chart they use, we should count the d3 libraries (38kB) in the bundle, leading to a total of 70kB

I'm not a huge fan of splitting them, because I see them as a single component (they do the same thing) with two variants.

I opened an issue for working on the topic of sparkling weight/perf: https://github.com/mui/mui-x/issues/9799

My idea is to keep this rough component as a quick win: "you need a sparkline, there is a simple one to use"
In a year if we confirm nobody asked for features I did not think about, we could go with a dedicated implementation (nearly as if it was an independent package) to get closer to [react-sparklines](https://github.com/borisyankov/react-sparklines) weight and perf

**Pull request merged**